ui/raidboss/emulator/overrides/RaidEmulatorOverlayApiHook.js

```diff
@@ -20,47 +41,79 @@ export default class RaidEmulatorOverlayApiHook {
     });
   }
 
-  call(msg) {
+  async call(msg) {
     if (msg.call === 'getCombatants') {
-      const tracker = this.emulator.currentEncounter.encounter.combatantTracker;
-      const timestamp = this.currentLogTime;
-      return new Promise((res) => {
-        const combatants = [];
-        const hasIds = msg.ids !== undefined && msg.ids.length > 0;
-        const hasNames = msg.names !== undefined && msg.names.length > 0;
+      return await this._getCombatantsOverride(msg);
+    } else if (cachedCactbotCalls.includes(msg.call)) {
+      if (!this.connected) {
+        if (this.hasCachedCactbotCall(msg))
+          return await this._cactbotCachedCall(msg);
```

Comment:
I was thinking of an icon top-right or something that shows connected/cached/disconnected, maybe with a hover for cached time. Or maybe if it's cached/disconnected, it should always show the modal, or show a different dialog?

Reply:
I think an icon would be good, because then if people send screenshots of raidemulator not working we can see it.

Reply:
Added a modal when opening raidemulator if disconnected, as well as connection indicator. The indicator has a tooltip as well.

![image](https://user-images.githubusercontent.com/6119598/121913777-ae6c6c00-ccff-11eb-99ba-d45ca132f869.png)
![image](https://user-images.githubusercontent.com/6119598/121913829-b9270100-ccff-11eb-95df-49f7e362a864.png)



![image](https://user-images.githubusercontent.com/6119598/121913903-ca700d80-ccff-11eb-8575-db2326e8c938.png)
![image](https://user-images.githubusercontent.com/6119598/121913950-d360df00-ccff-11eb-9ea4-7976897e89d1.png)



![image](https://user-images.githubusercontent.com/6119598/121914021-e5db1880-ccff-11eb-9464-586ad383ed53.png)

Reply:
Is there like a plug with a line through it for no, or something to visually distinguish them other than red/green colors?

Also sorry, but I still continue to think that caching will lead to confusion here and it'd be better to only support connected/unconnected states and not also cached (at some point).

Reply:
Cached is an addition with this PR, I can remove that if you're against it entirely.

The icons are just FontAwesome icons, I didn't see a good option for disconnected plug or anything. If you have a better suggestion for a connected and/or disconnected icon, it's easy enough to adjust.

https://fontawesome.com/v5.15/icons?d=gallery&p=2&m=free

Reply:
Yeah, sorry.  I am against it a little bit, mostly because I feel like cactbot gets enough support as it is, so keeping modes simpler for users goes a long way towards keeping cactbot more explainable.  If you feel like this would be really useful, I'm happy to hear you out?

Re: icons.  I could imagine `handshake` vs `handshake-slash`.  Alternatively `plug` vs `ban`?

Reply:
The biggest reason for caching is language settings, in my opinion. The rest are irrelevant.

Reply:
Gotcha!  I think in that case, I think it'd be better to figure out the language from the log file itself, like along the lines of https://github.com/quisquous/cactbot/pull/2942#issuecomment-851104263.  It'd be extremely convenient to be able to just point a user at the gh-pages version of raidemulator to test out a log file without having any other dependencies.   Having caching wouldn't help that use case.

I don't think that has to be done in this patch, but mostly trying to say "that's how I think this could be solved in the future".  This keeps "what language am I parsing in" independent of any ambiguous previous state and dependent on the log file itself.

Reply:
It already figures out parser language from the log file, the cached language settings are for trigger and timeline display. If caching is dropped and a CN user is using raidboss, it'll show triggers and timelines in chinese if they're connected to ACT and english if not.

Reply:
Oh, I see.  The display language, not the parsing language.  We could default to browser or parser language as the display language in those cases then?

Reply:
This javascript should detect the most appropriate language and fall back to 'en' by default, if someone with a non-'en' language could verify?
```js
[...navigator.languages, 'en']
  .map((l) => l.split('-')[0].toLowerCase())
  .filter((l) => ['en', 'de', 'fr', 'ja', 'cn', 'ko'].includes(l))[0];
```

Reply:
Alright, these changes are made. Please note if you're testing these that the timeline trigger bug is still present in this PR, so test against an encounter without timeline triggers or you're goona have a bad time.